node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+			// check if our knowledge of the peer already contains this assignment
+			match entry.known_by.entry(peer_id.clone()) {
+				hash_map::Entry::Occupied(knowledge) => {
+					if knowledge.get().known_messages.contains(&fingerprint) {
+						modify_reputation(ctx, peer_id, COST_DUPLICATE_MESSAGE).await;
+						return;
+					}
+				}
+				hash_map::Entry::Vacant(_) => {
+					modify_reputation(ctx, peer_id.clone(), COST_UNEXPECTED_MESSAGE).await;
+				}
+			}
+
+			// if the assignment is known to be valid, reward the peer
+			if entry.knowledge.known_messages.contains(&fingerprint) {
+				modify_reputation(ctx, peer_id.clone(), BENEFIT_VALID_MESSAGE).await;
+				entry.known_by.entry(peer_id).or_default().known_messages.insert(fingerprint.clone());
+				return;
+			}
+
+			// FIXME: possibly deadlocks due to https://github.com/paritytech/polkadot/issues/2149
+			// unless ApprovalVoting does not .await for the reply
+			let (tx, rx) = oneshot::channel();
+
+			ctx.send_message(AllMessages::ApprovalVoting(ApprovalVotingMessage::CheckAndImportAssignment(
+				assignment.clone(),
+				tx,
+			))).await;
+
+			let result = match rx.await {
+				Ok(result) => result,
+				Err(_) => {
+					tracing::debug!(
+						target: LOG_TARGET,
+						"The approval voting subsystem is down",
+					);
+					return;
+				}
+			};
+
+			match result {
+				AssignmentCheckResult::Accepted | AssignmentCheckResult::AcceptedDuplicate => {
+					if result == AssignmentCheckResult::Accepted {
+						modify_reputation(ctx, peer_id.clone(), BENEFIT_VALID_MESSAGE_FIRST).await;
+					}
+					entry.knowledge.known_messages.insert(fingerprint.clone());
+					entry.known_by
+						.entry(peer_id)
+						.or_default()
+						.known_messages
+						.insert(fingerprint.clone());
+				}
+				AssignmentCheckResult::TooFarInFuture => {
+					modify_reputation(ctx, peer_id, COST_ASSIGNMENT_TOO_FAR_IN_THE_FUTURE).await;
+					return;
+				}
+				AssignmentCheckResult::Bad => {
+					modify_reputation(ctx, peer_id, COST_INVALID_MESSAGE).await;
+					return;
+				}
+			}
+		} else {
+			entry.knowledge.known_messages.insert(fingerprint.clone());
+		}
+
+		match entry.candidates.get_mut(&claimed_candidate_index) {
+			Some(candidate_entry) => {
+				// set the approval state for validator_index to Assigned
+				// unless the approval state is set already
+				candidate_entry.approvals
+					.entry(validator_index)
+					.or_insert_with(|| ApprovalState::Assigned(assignment.cert.clone()));
+			}
+			None => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					hash = ?block_hash,
+					?claimed_candidate_index,
+					"Expected a candidate entry on import_and_circulate_assignment",
+				);
+			}
+		}
+
+		// Dispatch a ApprovalDistributionV1Message::Assignment(assignment, candidate_index)
+		// to all peers in the BlockEntry's known_by set,
+		// excluding the peer in the source, if source has kind MessageSource::Peer.
+		let maybe_peer_id = source.peer_id();
+		let peers = self.peer_views
```

Comment:
I think the guide is wrong here and actually we should only send to peers who don't have the fingerprint in `known_by`. That could be combined with the step below of adding the fingerprint of hte assignment to the knowledge of the peer

Reply:
Alternatively, the same result can be achieved by returning if the result is `AcceptedDuplicate` above or if `entry.knowledge` contains the fingerprint already.

Reply:
Why trying to write a test for this, I realized I don't understand how `AcceptedDuplicate` could be even triggered.
If we already know about an assignment that would trigger early return with `BENEFIT_VALID_MESSAGE`, unless the `source` is `Local` and in that case I think we should also early return (we currently don't). 

It seems to me that at this point (in this function) nobody except for the source knows about the assignment or am I missing something?
Because we only add a fingerprint to a peer's knowledge after we add it to ours, no? And once we remove it from our knowledge, we won't add it back. 
Same applies to the approvals.

Reply:
I'm trying to recall exactly what I was thinking about at the time.

One thing that might be helpful to understand is that "duplicate" assignments aren't necessarily equal. There is more than one way each validator can be assigned to each core. However, I think you're right because the fingerprints will be equal since we don't account for delay-tranche. I think that's alright, though, no need to change that.

I think we can ignore `AcceptedDuplicate` because it'll be checked earlier. We should just use it to avoid giving a benefit. However, I think we need to keep `AcceptedDuplicate` in the interface of `ApprovalVoting` even though it isn't used here.